types/coin.go

```diff
@@ -101,127 +132,166 @@ func (coins Coins) String() string {
 func (coins Coins) Plus(coinsB Coins) Coins {
+	return coins.safePlus(coinsB)
+}
+
+// safePlus will perform addition of two coins sets. If both coin sets are
+// empty, then an empty set is returned. If only a single set is empty, the
+// other set is returned. Otherwise, the coins are compared in order of their
+// denomination and addition only occurs when the denominations match, otherwise
+// the coin is simply added to the sum assuming it's not zero.
+func (coins Coins) safePlus(coinsB Coins) Coins {
 	sum := ([]Coin)(nil)
 	indexA, indexB := 0, 0
 	lenA, lenB := len(coins), len(coinsB)
+
 	for {
 		if indexA == lenA {
 			if indexB == lenB {
+				// return nil coins if both sets are empty
 				return sum
 			}
-			return append(sum, coinsB[indexB:]...)
+
+			// return set B (excluding zero coins) if set A is empty
+			return append(sum, removeZeroCoins(coinsB[indexB:])...)
 		} else if indexB == lenB {
-			return append(sum, coins[indexA:]...)
+			// return set A (excluding zero coins) if set B is empty
+			return append(sum, removeZeroCoins(coins[indexA:])...)
 		}
+
 		coinA, coinB := coins[indexA], coinsB[indexB]
+
 		switch strings.Compare(coinA.Denom, coinB.Denom) {
-		case -1:
+		case -1: // coin A denom < coin B denom
 			if coinA.IsZero() {
-				// ignore 0 sum coin type
+				// ignore 0 sum coin
 			} else {
 				sum = append(sum, coinA)
 			}
+
 			indexA++
-		case 0:
-			if coinA.Amount.Add(coinB.Amount).IsZero() {
+
+		case 0: // coin A denom == coin B denom
+			res := coinA.Plus(coinB)
+			if res.IsZero() {
 				// ignore 0 sum coin type
 			} else {
-				sum = append(sum, coinA.Plus(coinB))
+				sum = append(sum, res)
 			}
+
 			indexA++
 			indexB++
-		case 1:
+
+		case 1: // coin A denom > coin B denom
 			if coinB.IsZero() {
-				// ignore 0 sum coin type
+				// ignore 0 sum coin
 			} else {
 				sum = append(sum, coinB)
 			}
+
 			indexB++
 		}
 	}
 }
 
-// Negative returns a set of coins with all amount negative
-func (coins Coins) Negative() Coins {
-	res := make([]Coin, 0, len(coins))
-	for _, coin := range coins {
-		res = append(res, Coin{
-			Denom:  coin.Denom,
-			Amount: coin.Amount.Neg(),
-		})
+// Minus subtracts a set of coins from another.
+//
+// e.g.
+// {2A, 3B} - {A} = {A, 3B}
+// {2A} - {0B} = {2A}
+// {A, B} - {A} = {B}
+//
+// CONTRACT: Minus will never return Coins where one Coin has a non-positive
+// amount. In otherwords, IsValid will always return true.
+func (coins Coins) Minus(coinsB Coins) Coins {
+	diff, hasNeg := coins.SafeMinus(coinsB)
+	if hasNeg {
+		panic("negative coin amount")
 	}
-	return res
+
+	return diff
 }
 
-// Minus subtracts a set of coins from another (adds the inverse)
-func (coins Coins) Minus(coinsB Coins) Coins {
-	return coins.Plus(coinsB.Negative())
+// SafeMinus performs the same arithmetic as Minus but returns a boolean if any
+// negative coin amount was returned.
+func (coins Coins) SafeMinus(coinsB Coins) (Coins, bool) {
```

Comment:
why is `SafeMinus` exportable at all, do any other modules require it?

Reply:
@fedekunze because `safePlus` Is not needed really outside the scope of the package (assuming you're working with positive numbers -- which you should be). i.e. you'll never do `SafePlus({"A", 1}, {"A", -5})`.

@cwgoes -- The reasoning behind `SafeMinus` is fee deduction in the ante handler where we need to check if an account has enough funds and a similar case in the bank keeper (`subtractCoins`). I suppose we can just use `IsAllLTE` here?

Reply:
SafeMinus seems fine.

Reply:
Thanks @alexanderbez, that seems reasonable, the comment is clear.